src/register/macros.rs

```diff
@@ -1,255 +1,198 @@
-        unsafe fn _write(bits: usize) {
-            match () {
-                #[cfg(riscv)]
-                () => core::arch::asm!(concat!("csrrw x0, ", stringify!($csr_number), ", {0}"), in(reg) bits),
-
-                #[cfg(not(riscv))]
-                () => unimplemented!(),
-            }
-        }
-    };
-}
+/// Generates the imports and types that all CSRs need
+macro_rules! csr_boilerplate {
+    ($name:ident, $type:ty, $size:ty) => {
+        paste::paste! {
+            use core::arch::asm;
 
-macro_rules! write_csr_rv32 {
-    ($csr_number:literal) => {
-        /// Writes the CSR
-        #[inline]
-        #[allow(unused_variables)]
-        unsafe fn _write(bits: usize) {
-            match () {
-                #[cfg(riscv32)]
-                () => core::arch::asm!(concat!("csrrw x0, ", stringify!($csr_number), ", {0}"), in(reg) bits),
-
-                #[cfg(not(riscv32))]
-                () => unimplemented!(),
-            }
-        }
-    };
-}
+            use tock_registers::LocalRegisterCopy;
+            use tock_registers::register_bitfields;
 
-macro_rules! write_csr_as_usize {
-    ($csr_number:literal) => {
-        write_csr!($csr_number);
+            /// This allows the enums to be accessed directly from the module's
+            /// namespace instead of having to go through the bitfield's as well
+            pub use [< $name:lower >]::*;
 
-        /// Writes the CSR
-        #[inline]
-        pub fn write(bits: usize) {
-            unsafe { _write(bits) }
-        }
-    };
-}
+            /// An alias of the CSR's index
+            const INDEX: u16 = $crate::register::addresses::[< CSR_ $name:upper >];
 
-macro_rules! write_csr_as_usize_rv32 {
-    ($csr_number:literal) => {
-        write_csr_rv32!($csr_number);
+            /// A typedef of the [`tock_registers::RegisterLongName`] associated with this CSR
+            type LongName = [< $name:lower >]::Register;
 
-        /// Writes the CSR
-        #[inline]
-        pub fn write(bits: usize) {
-            unsafe { _write(bits) }
-        }
-    };
-}
+            /// A typedef of the [`Field`] associated with this CSR
+            pub type Field = tock_registers::fields::Field<$size, LongName>;
 
-macro_rules! set {
-    ($csr_number:literal) => {
-        /// Set the CSR
-        #[inline]
-        #[allow(unused_variables)]
-        unsafe fn _set(bits: usize) {
-            match () {
-                #[cfg(riscv)]
-                () => core::arch::asm!(concat!("csrrs x0, ", stringify!($csr_number), ", {0}"), in(reg) bits),
-
-                #[cfg(not(riscv))]
-                () => unimplemented!(),
-            }
+            /// A typedef of the [`LocalRegisterCopy`] associated with this CSR
+            pub type Local = LocalRegisterCopy<$size, LongName>;
         }
     };
 }
 
-macro_rules! clear {
-    ($csr_number:literal) => {
-        /// Clear the CSR
+/// Generates the read functions for a CSR
+macro_rules! csr_reads {
```

Comment:
Looks like we are generating these read, write and other functions for each csr that is defined. Why not define a CSR type and provide implementations for that type instead? I am just looking at the example of tock and trying to understand why this design was chosen as opposed to what tock does. [[Reference](https://github.com/tock/tock/blob/9cc3d42572ce1ad117895ad81767f671e906d3a6/libraries/riscv-csr/src/csr.rs#L108)]

Reply:
I was mostly looking at how [tock instantiated the CSRs](https://github.com/tock/tock/blob/9cc3d42572ce1ad117895ad81767f671e906d3a6/arch/riscv/src/csr/mod.rs#L159) and thinking along the same lines. Salus seems to have [inherited the same design](https://github.com/rivosinc/salus/blob/main/riscv-regs/src/csrs/defs.rs). I think this kind of falls into the first category you defined. And it looks like the inconsistent access you described was solved by Tock by implementing functions directly as part of the CSR type e.g. https://github.com/tock/tock/blob/9cc3d42572ce1ad117895ad81767f671e906d3a6/arch/riscv/src/csr/mod.rs#L315

Since I haven't played with either of the interfaces much, I don't have a strong opinion on which way to go. It was more of a curiosity to understand why we are choosing this design. 

The way it was written in tock/salus by introducing a generic type for RW-CSR ([ReadWriteRiscvCsr](https://github.com/tock/tock/blob/9cc3d42572ce1ad117895ad81767f671e906d3a6/arch/riscv/src/csr/mod.rs#L41)) and an interface for it and then using that to define a struct [CSR](https://github.com/tock/tock/blob/9cc3d42572ce1ad117895ad81767f671e906d3a6/arch/riscv/src/csr/mod.rs#L41) for all CSRs in RV seemed like a fine separation. But as you pointed out, there might be other limitations/drawbacks of this. Maybe @abrestic-rivos has more thoughts on this based on his usage of the tock registers.

Reply:
I should also note that the first category makes it difficult to implement CSRs with layouts similar to the PMP configuration registers (The same pattern repeated). You can see [here](https://github.com/tock/tock/blob/9cc3d42572ce1ad117895ad81767f671e906d3a6/arch/riscv/src/csr/pmpconfig.rs#L6-L50) the impact of doing it this way. The bitfields are far more convoluted because of this setup. 

> And it looks like the inconsistent access you described was solved by Tock by implementing functions directly as part of the CSR type

The problem with this is that _all_ functions are in the same namespace, so it's not as easy to add CSR-specific functions for each CSR without the names getting too convoluted.

While I'm definitely in favor of the approach I've taken here, if there are notable benefits for the other method I don't mind switching.

Reply:
I understand your point about having to define all the functions within the same namespace. That was one of the things that bothered me as well about the approach. 

Anyways, I didn't mean to question your approach. I just wanted to understand your thoughts and reasoning behind choosing the approach.  I trust your judgement and evaluation and we can go ahead with that. Thanks for the details! :)

Reply:
> The way it was written in tock/salus by introducing a generic type for RW-CSR ([ReadWriteRiscvCsr](https://github.com/tock/tock/blob/9cc3d42572ce1ad117895ad81767f671e906d3a6/arch/riscv/src/csr/mod.rs#L41)) and an interface for it and then using that to define a struct [CSR](https://github.com/tock/tock/blob/9cc3d42572ce1ad117895ad81767f671e906d3a6/arch/riscv/src/csr/mod.rs#L41) for all CSRs in RV seemed like a fine separation. But as you pointed out, there might be other limitations/drawbacks of this. Maybe @abrestic-rivos has more thoughts on this based on his usage of the tock registers.

FWIW, this works pretty well for us (with one exception[^0]) so if the eventual goal is for Salus to use this too, I'd like to see a compelling reason for why we should change it :)

The CSR-specific-methods issue can be handled by defining and implementing traits for `ReadWriteRiscvCsr<CSR>` or `LocalRegisterCopy<CSR>`. For example, if you wanted to define a `cause()` method like you've done in the subsequent patch, you could do something like:

```
trait XCauseHelpers {
    fn cause(&self) -> Trap;

    fn is_interrupt(&self) -> bool {
        matches!(self.cause(), Trap::Interrupt(_))
    }
}

trait XCause: LongRegisterName {}
impl XCause for mcause::Register {}
impl XCause for scause::Register {}

impl<R: XCause> XCauseHelpers for ReadWriteRiscvCsr<usize, R> {
    fn cause(&self) -> Trap { ... }
}
```

And then you'd use it like this:

```
let trap = CSR.mcause.cause();
```

That doesn't seem any less ergonomic that what you're proposing, and involves less use of macros (which IMO are less readable than traits/generics).

[^0]: The tock `Readable`/`Writable` traits aren't object safe, so things get annoying if you want to create trait objects from CSRs, which is useful when dealing with a range of identically-defined CSRs (think `hpmcounter`).

Reply:
> ...  if the eventual goal is for Salus to use this too, I'd like to see a compelling reason for why we should change it :)

I would definitely like this PR to end up producing the one crate that rules them all, regardless of if it's what I've proposed here or something else.

> The CSR-specific-methods issue can be handled by defining and implementing traits for ReadWriteRiscvCsr<CSR> or LocalRegisterCopy<CSR>. For example, if you wanted to define a cause() method like you've done in the subsequent patch, you could do something like:
>```
>trait XCauseHelpers {
>    fn cause(&self) -> Trap;
>
>    fn is_interrupt(&self) -> bool {
>        matches!(self.cause(), Trap::Interrupt(_))
>    }
>}
>
>trait XCause: LongRegisterName {}
>impl XCause for mcause::Register {}
>impl XCause for scause::Register {}
>
>impl<R: XCause> XCauseHelpers for ReadWriteRiscvCsr<usize, R> {
>    fn cause(&self) -> Trap { ... }
>}
>```
> And then you'd use it like this:
> ```
> let trap = CSR.mcause.cause();
> ```
> That doesn't seem any less ergonomic that what you're proposing, and involves less use of macros (which IMO are less readable than traits/generics).

I didn't think of that. This makes a struct/trait based approach a lot more palatable (although I still like modules more :)), especially because we can reuse some traits between CSRs for different privilege levels. 

> The tock Readable/Writable traits aren't object safe, so things get annoying if you want to create trait objects from CSRs, which is useful when dealing with a range of identically-defined CSRs (think hpmcounter)

This was another issue I had too. Having to make the array for HPM counters `dyn` ([here](https://github.com/rivosinc/salus/blob/main/riscv-regs/src/csrs/mod.rs#L76)) is not ideal (I believe this requires a vtable lookup?). I think that the easiest way around this is to instead provide [indexed functions](https://github.com/rivosinc/riscv/blob/6ff7db4296d72b3c046b2b02079dd0e4dfd73328/src/register/pmpcfgx.rs#L152-L177), where the caller just provides an integer for the CSR they want and the function uses a `match` to call the correct function. As long as the function calls in the `match` are inlined this should basically become a jump table.

My one remaining issue, however, is with CSRs like the PMPs. If we wanted to use something like `ReadWriteRiscvCsr` for everything, the PMP created by `register_bitfields!` would have to look like [this](https://github.com/tock/tock/blob/9cc3d42572ce1ad117895ad81767f671e906d3a6/arch/riscv/src/csr/pmpconfig.rs#L6-L50), which is terrible. That being said, in 99% of cases the user will likely want to use an indexed function for CSRs like this so they don't have to be exposed to this interface directly. However, that still means the internals aren't going to be very pretty. An easy way around this would be to implement PMPs as a separate type with the exact same functions, so the user wouldn't be aware of the difference without looking at the code/docs. This would only cause an issue if the user wanted to implement some sort of generic function using a `ReadWriteRiscvCsr`, but that's not something my design makes possible either so I don't think it's a big deal.

Any other thoughts? With what Andrew's pointed out here I'm more willing to go with a struct/trait based setup than before, although I still think a module-based structure makes more sense for CSRs given that they're effectively singletons.

Reply:
> This was another issue I had too. Having to make the array for HPM counters dyn ([here](https://github.com/rivosinc/salus/blob/main/riscv-regs/src/csrs/mod.rs#L76)) is not ideal (I believe this requires a vtable lookup?). I think that the easiest way around this is to instead provide [indexed functions](https://github.com/rivosinc/riscv/blob/6ff7db4296d72b3c046b2b02079dd0e4dfd73328/src/register/pmpcfgx.rs#L152-L177), where the caller just provides an integer for the CSR they want and the function uses a match to call the correct function. As long as the function calls in the match are inlined this should basically become a jump table.

Yeah, it's a vtable lookup. There's definitely room to improve how we've handled the counter CSRs. I like the indexed approach if we care about the generated assembly for the particular CSRs. (It is effectively what Linux does for the counter CSRs, via a mess of cpp macros).

> My one remaining issue, however, is with CSRs like the PMPs.

Yes, that's a bit of a mess; we don't have the PMP CSRs in Salus so I've ignored them :) 

Given that it's not a pattern that's oft repeated in CSR definitions (CSR divided into independent, identical sub-registers), I would be tempted to skip the `register_bitfields!` and implement setting of an individual PMP entry via a helper trait.

Reply:
> Yeah, it's a vtable lookup.

Definitely not something we want on firmware, but I think the indexed approach is acceptable for us. 

> Yes, that's a bit of a mess; we don't have the PMP CSRs in Salus so I've ignored them :)

Can't blame you there.

> Given that it's not a pattern that's oft repeated in CSR definitions (CSR divided into independent, identical sub-registers), I would be tempted to skip the register_bitfields! and implement setting of an individual PMP entry via a helper trait.

I think that's fine. I'll look further into a design that uses traits; I had already put some stuff together but got fed up with the problems I had initially and decided to switch. I'll probably push that out as a separate PR so we can review both separately.